### ISSUE TITLE
Evaluation fails when order of labels is different

### ISSUE THREAD

**Issue by joostdewit** [deeplearning4j/DataVec#105]
As a data scientist I'm trying to train a classifier to recognize single digits in images. For this I created a synthetic dataset containing thousands of labeled examples. 

During training I split the data in a training and test set. The training step results in a low training loss, but evaluating the model shows an accuracy of 0%.

The training and test iters are created as follows:

```
ParentPathLabelGenerator labelMaker = new ParentPathLabelGenerator();

FileSplit filesInDir = new FileSplit(dataDir, allowedExtensions, randNumGen);
BalancedPathFilter pathFilter = new BalancedPathFilter(randNumGen, allowedExtensions, labelMaker, 0, 0, 10, 2000, labelWhiteList.toArray(new String[labelWhiteList.size()]));

ImageRecordReader trainRecordReader = new ImageRecordReader(height, width, channels, labelMaker);
trainRecordReader.initialize(filesInDirSplit[0]);

ImageRecordReader testRecordReader = new ImageRecordReader(height, width, channels, labelMaker);
testRecordReader.initialize(filesInDirSplit[1]);

DataSetIterator trainIter = new RecordReaderDataSetIterator(trainRecordReader, 10);
DataSetIterator testIter = new RecordReaderDataSetIterator(testRecordReader, 10);
```

Evaluation is done in every epoch as follows:

```
for (int i = 0; i < NUM_OF_EPOCHS; i++) {
	// Fit
	cnn.fit(trainIter);
	// Evaluate
	Evaluation eval = new Evaluation(outputNum);
	while (testIter.hasNext()) {
		DataSet ds = testIter.next();
		INDArray output = cnn.output(ds.getFeatureMatrix(), false);
		eval.eval(ds.getLabels(), output);
	}
	log.info(eval.stats());
	testIter.reset();
}
```

It appears that the order of the labels differs for both DatasetIterators:

```
trainIter's labels : [3, 2, 1, 0, 7, 6, 5, 4, 9, 8]
testIter's labels  : [5, 9, 3, 4, 0, 6, 1, 2, 8, 7]
```

The confusion matrix shows that the predictions are consistently wrong. A '3' is mostly recognized as a '5' for example. I suspect that the difference in label ordering is causing the evaluation to fail.

**Comment by AlexDBlack** [comment]
Fixed: https://github.com/deeplearning4j/DataVec/pull/106

That was easier to reproduce than expected.
As a workaround until next release: trainRecordReader.setLabels(testRecordReader.getLabels())
or Collections.sort(xRecordReader.getLabels()) on both readers should also work.

**Comment by joostdewit** [comment]
Sorting just the labels doesn't seem to solve the issue. I'm now training 27 classes [a .. z, no-char].

My code looks like this:

```
// Loading the data
FileSplit filesInDir = new FileSplit(dataDir, allowedExtensions, randNumGen);
BalancedPathFilter pathFilter = new BalancedPathFilter(randNumGen, allowedExtensions, labelMaker, 0, 0, 1000, 2500,  [a .. z, no-char]);

// Split the data
InputSplit[] filesInDirSplit = filesInDir.sample(pathFilter, 80, 20);
log.info("Training set size is {}", filesInDirSplit[0].length());
log.info("Test set size is {}", filesInDirSplit[1].length());

// Readers
ImageRecordReader trainRecordReader = createImageRecordReader(filesInDirSplit[0], height, width, channels);
ImageRecordReader testRecordReader = createImageRecordReader(filesInDirSplit[1], height, width, channels);

// Sort labels
Collections.sort(trainRecordReader.getLabels());
Collections.sort(testRecordReader.getLabels());

// Dataset Iterators
DataSetIterator trainIter = new RecordReaderDataSetIterator(trainRecordReader, BATCH_SIZE, 1, outputNum);
DataSetIterator testIter = new RecordReaderDataSetIterator(testRecordReader, BATCH_SIZE, 1, outputNum);
```

For each epoch, evaluation is done as follows.
```
log.info("Evaluate model....");
log.info("Train iter....");
for (int ci = 0; ci < trainIter.getLabels().size(); ci++) {
	log.info("Class {} has label {}", ci, trainIter.getLabels().get(ci));
}
log.info("Test iter....");
for (int ci = 0; ci < testIter.getLabels().size(); ci++) {
	log.info("Class {} has label {}", ci, testIter.getLabels().get(ci));
}
Evaluation eval = new Evaluation(outputNum);
while (testIter.hasNext()) {
        DataSet ds = testIter.next();
        INDArray output = cnn.output(ds.getFeatureMatrix(), false);
        eval.eval(ds.getLabels(), output);
}
```

This results in the following logs:
```
t.m.d.o.SingleCharacterRecognitionTrainer - Evaluate model....
t.m.d.o.SingleCharacterRecognitionTrainer - Train iter....
t.m.d.o.SingleCharacterRecognitionTrainer - Class 0 has label a
t.m.d.o.SingleCharacterRecognitionTrainer - Class 1 has label b
t.m.d.o.SingleCharacterRecognitionTrainer - Class 2 has label c
t.m.d.o.SingleCharacterRecognitionTrainer - Class 3 has label d
t.m.d.o.SingleCharacterRecognitionTrainer - Class 4 has label e
t.m.d.o.SingleCharacterRecognitionTrainer - Class 5 has label f
t.m.d.o.SingleCharacterRecognitionTrainer - Class 6 has label g
t.m.d.o.SingleCharacterRecognitionTrainer - Class 7 has label h
t.m.d.o.SingleCharacterRecognitionTrainer - Class 8 has label i
t.m.d.o.SingleCharacterRecognitionTrainer - Class 9 has label j
t.m.d.o.SingleCharacterRecognitionTrainer - Class 10 has label k
t.m.d.o.SingleCharacterRecognitionTrainer - Class 11 has label l
t.m.d.o.SingleCharacterRecognitionTrainer - Class 12 has label m
t.m.d.o.SingleCharacterRecognitionTrainer - Class 13 has label n
t.m.d.o.SingleCharacterRecognitionTrainer - Class 14 has label no-char
t.m.d.o.SingleCharacterRecognitionTrainer - Class 15 has label o
t.m.d.o.SingleCharacterRecognitionTrainer - Class 16 has label p
t.m.d.o.SingleCharacterRecognitionTrainer - Class 17 has label q
t.m.d.o.SingleCharacterRecognitionTrainer - Class 18 has label r
t.m.d.o.SingleCharacterRecognitionTrainer - Class 19 has label s
t.m.d.o.SingleCharacterRecognitionTrainer - Class 20 has label t
t.m.d.o.SingleCharacterRecognitionTrainer - Class 21 has label u
t.m.d.o.SingleCharacterRecognitionTrainer - Class 22 has label v
t.m.d.o.SingleCharacterRecognitionTrainer - Class 23 has label w
t.m.d.o.SingleCharacterRecognitionTrainer - Class 24 has label x
t.m.d.o.SingleCharacterRecognitionTrainer - Class 25 has label y
t.m.d.o.SingleCharacterRecognitionTrainer - Class 26 has label z
t.m.d.o.SingleCharacterRecognitionTrainer - Test iter....
t.m.d.o.SingleCharacterRecognitionTrainer - Class 0 has label b
t.m.d.o.SingleCharacterRecognitionTrainer - Class 1 has label c
t.m.d.o.SingleCharacterRecognitionTrainer - Class 2 has label f
t.m.d.o.SingleCharacterRecognitionTrainer - Class 3 has label h
t.m.d.o.SingleCharacterRecognitionTrainer - Class 4 has label j
t.m.d.o.SingleCharacterRecognitionTrainer - Class 5 has label m
t.m.d.o.SingleCharacterRecognitionTrainer - Class 6 has label no-char
t.m.d.o.SingleCharacterRecognitionTrainer - Class 7 has label p
t.m.d.o.SingleCharacterRecognitionTrainer - Class 8 has label q
t.m.d.o.SingleCharacterRecognitionTrainer - Class 9 has label u
t.m.d.o.SingleCharacterRecognitionTrainer - Class 10 has label v
t.m.d.o.SingleCharacterRecognitionTrainer - Class 11 has label w
t.m.d.o.SingleCharacterRecognitionTrainer - Class 12 has label x
t.m.d.o.SingleCharacterRecognitionTrainer - Class 13 has label y
t.m.d.o.SingleCharacterRecognitionTrainer - Class 14 has label z
```

In my opinion this shows that the issue isn't solved yet. Which makes sense, since only the labels are sorted. The data isn't ordered accordingly. Furthermore, some labels appear to be missing in the `testiter`.

**Comment by saudet** [comment]
If the test set doesn't contain samples for all labels, then its labels shouldn't be used. Just use the set collected from the training set.

**Comment by joostdewit** [comment]
It would be nice when the `Evaluation` class can cope with missing labels. Can you please confirm that sorting the labels isn't screwing up the link to the data columns/rows?

**Comment by AlexDBlack** [comment]
The problem isn't the Evaluation class - it's in how the ImageRecordReader assigns labels.
It looks at all images in the train/test set, and collects the set of possible labels from that.
As @saudet said, you shouldn't be using the labels from the test set if it isn't complete: the ImageRecordReader in that case has no way to know about the missing labels - unless you explicitly set them.

testRecordReader.setLabels(trainRecordReader.getLabels())
